No CORS headers sent if Exception is thrown

Lavinski opened the pull request:
Normally headers are added however if a controller throws an exception then CORS headers not be present.

All matching responses should have CORS headers added. This problem affects the use of middleware (including the UseExceptionHandler middleware).

This PR fixes the issue by moving the adding of the headers to `Response.OnStarting`.

https://github.com/aspnet/Home/issues/2378
https://github.com/aspnet/Home/issues/3220

dnfclas commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/aspnet/CORS?pullRequest=163) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/aspnet/CORS?pullRequest=163) before we can accept your contribution.<br/><br/>:x: Lavinski [sign now](https://cla.dotnetfoundation.org/aspnet/CORS?pullRequest=163)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/aspnet/CORS?pullRequest=163) it.</sub>

Lavinski commented:
This feature is ready from a development position. I don't foresee any breaking changes and this is a highly requested feature. Ready for review.

rynowak commented:
@Tratcher - do you have thoughts about this? I'm deeply passionate about us fixing this issue, we get this feedback a lot.

The change itself is designed to make an end-run around `HttpResponse.Clear()` - I'm not aware of anywhere else we do this. Funnily enough to the exception handling middleware sets cache heads to prevent caching on failures (developer exception page does not). Other there other things that need to work the same way as CORS (CSP for example) or is CORS special?

rynowak commented:
Note that this is also duplicated in MVC - and it looks like it would need the same changes there. https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Cors/CorsAuthorizationFilter.cs

Tratcher commented on src/Microsoft.AspNetCore.Cors/Infrastructure/CorsMiddleware.cs:
```diff
@@ -106,26 +106,38 @@ public class CorsMiddleware
+                    else
+                    {
+                        context.Response.OnStarting(_ => {
+                            ApplyCorsHeaders(context, corsPolicy);
```
Catch all exceptions in this callback. An exception on this call path would bypass all exception pages and result in a blank 500.

Lavinski replied:
This is middleware which means it can be used without mvc. Furthermore exceptions on this path already result in a blank 500 except it won’t have CORS headers.

You should use the exception handler middleware to provide consistent, last chance errors to the client. This will also help in cases were mvc is used with other serving middleware.

Tratcher replied:
It's not practical for the exception handling middleware to handle exceptions in OnStarting, those callbacks are invoked directly by the server.

Lavinski replied:
I’m sorry I think you’ve misunderstood (or I’m missing something). There is no exception handling in OnStarting. 

This change is primarily to ensure CORS middleware works with ‘UseExceptionHandler’ but applies to all similar cases.

The flow goes like this:
1. An exception is uncaught in an API controller.
2. Exception is handled by global exception handler.
3. CORS headers are added to error details can be used by the client.

Is there anything else I can clarify?

Lavinski replied:
@Tratcher or did you mean wrapping `ApplyCorsHeaders` in a `try catch` in case of an error in the callback?
```
context.Response.OnStarting(_ => {
    ApplyCorsHeaders(context, corsPolicy);
```

Tratcher replied:
Right. We want to avoid throwing exceptions in the OnStarting code path as it has a worse error handling experience. These wouldn't be app exceptions, but exceptions from within CORS itself. I'm not aware of any, but we'd rather not risk it.

JulijaRamoskiene commented:
I think it is a mistake to have duplicated code in CorsAuthorizationFilter and CorsMiddleware E.g.: preflight requests are now handled differently between them. CorsAuthorizationFilter sets 200 OK response code and CorsMiddleware sets 204 No Content.

JulijaRamoskiene commented:
Since CorsAuthorizationFilter and CorsMiddleware both call ApplyResult () on ICorsService, maybe it would make sense to move the fix to that method?

Lavinski commented:
@Tratcher I've added a `try` for code inside `OnStarting` and I've changed the `OnStarting` to pass through the context (as opposed to using the variable via a closure).

Tratcher commented on src/Microsoft.AspNetCore.Cors/Infrastructure/CorsMiddleware.cs:
```diff
@@ -18,18 +19,21 @@ public class CorsMiddleware
-            ICorsPolicyProvider policyProvider)
-            : this(next, corsService, policyProvider, policyName: null)
+            ICorsPolicyProvider policyProvider,
+            ILoggerFactory loggerFactory)
```
AppVeyor is warning about API signature breaks here.
https://ci.appveyor.com/project/aspnetci/cors/build/1.0.476

For this 2.2 milestones we cannot change any existing APIs, only add new overloads.

Tratcher replied:
See what I did over here:
https://github.com/aspnet/BasicMiddleware/pull/340/files

Tratcher commented on src/Microsoft.AspNetCore.Cors/Infrastructure/CorsMiddleware.cs:
```diff
@@ -106,26 +125,45 @@ public class CorsMiddleware
+                        context.Response.OnStarting(c => {
+                            try
+                            {
+                                ApplyCorsHeaders((HttpContext)c, corsPolicy);
```
Pass both parameters in via state like this: https://github.com/aspnet/KestrelHttpServer/pull/2670/files#diff-4ab77515238e5e964ae9a3a4615205b1R713
Then this callback can become a cached instance.

pranavkm commented:
Merged via https://github.com/aspnet/CORS/commit/554855cab34961a27a6cf248fbb847b9dd8bd8d4. Thanks for the PR!
